### PR TITLE
Stream read pool and default s3 timeouts tuning

### PR DESCRIPTION
**Pull request by vikasvb90**
### Description
On upload of large shards (large segments or more number of smaller segments), we noticed s3 errors. For large segments, api timeouts were observed and for large shards of small segments connection acquisition timeout was observed as below. Also, for small instances having low cpu cores, stream read pool capacity becomes low which processes upload requests slowly.
This change is to increase the capacity of stream read pool and increase connection acquisition and api timeouts.

```
[2023-10-25T07:47:37,139][ERROR][o.o.i.s.RemoteDirectory  ] [51844e1377c6a637328871dc96b00d5e] Failed to upload blob _3h_Lucene90_0.tim
software.amazon.awssdk.core.exception.SdkClientException: Failed to send multipart upload requests.
        at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:111)
        at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:47)
        at org.opensearch.repositories.s3.async.AsyncTransferManager.handleException(AsyncTransferManager.java:278)
        at org.opensearch.repositories.s3.async.AsyncTransferManager.lambda$handleExceptionOrResponse$14(AsyncTransferManager.java:227)
        at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
        at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
        at software.amazon.awssdk.utils.CompletableFutureUtils.lambda$forwardExceptionTo$0(CompletableFutureUtils.java:79)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
```

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] ~New functionality includes testing.~
  - [ ] ~All tests pass~
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [ ] ~Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~
- [x] Commits are signed per the DCO using --signoff
- [ ] ~Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))~
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3ClientSettings.java [hunk 1]
```diff
@@ -205,7 +205,7 @@ final class S3ClientSettings {
     static final Setting.AffixSetting<TimeValue> CONNECTION_ACQUISITION_TIMEOUT = Setting.affixKeySetting(
         PREFIX,
         "connection_acquisition_timeout",
-        key -> Setting.timeSetting(key, TimeValue.timeValueMinutes(2), Property.NodeScope)
+        key -> Setting.timeSetting(key, TimeValue.timeValueMinutes(20), Property.NodeScope)
```
Comment (kkmr) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3ClientSettings.java:208]: Maybe add some comments on why 20 minutes is a good timeout?  Also, maybe make it a constant.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java [hunk 2]
```diff
@@ -110,7 +110,9 @@ public List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
         executorBuilders.add(new ScalingExecutorBuilder(PRIORITY_STREAM_READER, 1, halfProcMaxAt5, TimeValue.timeValueMinutes(5)));
 
         executorBuilders.add(new FixedExecutorBuilder(settings, FUTURE_COMPLETION, normalPoolCount(settings), 10_000, FUTURE_COMPLETION));
-        executorBuilders.add(new ScalingExecutorBuilder(STREAM_READER, 1, halfProcMaxAt5, TimeValue.timeValueMinutes(5)));
+        executorBuilders.add(
+            new ScalingExecutorBuilder(STREAM_READER, 1, 3 * allocatedProcessors(settings), TimeValue.timeValueMinutes(5))
```
Comment (kkmr) [inline plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java:114]: Add some comment here as well.

**Comment by kkmr** [comment]
Also, do we have any perf results showing how things have improved?

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3ClientSettings.java (hunk 1 above)
Comment (vikasvb90) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3ClientSettings.java:208]: It shouldn't be a constant. This is a repository setting and users should be able to override it. There's no one specific good configuration. This is just to handle upload bursts in special cases like force merge temporarily.

**Comment by vikasvb90** [comment]
> Also, do we have any perf results showing how things have improved?

This is not about perf. This is to handle upload bursts occurring due to large file uploads. Increasing timeouts is a temporary change till we use a separate dedicated slow client which can handle special cases by putting them behind a producer-consumer queue and processing them slowly without impacting critical uploads.

File: plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java (hunk 2 above)
Comment (vikasvb90) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java:114]: This is just an increase in capacity. What do you want me to put?

Comment (kkmr) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java:114]: Explain why you think 3*allocatedProcessors is good?

Comment (vikasvb90) [reply plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3RepositoryPlugin.java:114]: The explanation is that there are disk reads i.e IO happening in stream reader pool. This explanation will not hold true with the optimization planned https://github.com/opensearch-project/OpenSearch/issues/11263

**Review by gbbafna — APPROVED** [review]

**Review by kkmr — APPROVED** [review]

**Pull request merged**